Android: Enable views to be nested within <Text>

rigdern opened the pull request:
Potential breaking change: The signature of ReactShadowNode's onBeforeLayout method was changed
- Before: public void onBeforeLayout()
- After:  public void onBeforeLayout(NativeViewHierarchyOptimizer nativeViewHierarchyOptimizer)

Depends on this css-layout PR:
https://github.com/facebook/css-layout/pull/202

Implements same feature as this iOS PR:
https://github.com/facebook/react-native/pull/7304

Previously, only Text and Image could be nested within Text. Now, any view can be nested within Text. One restriction of this feature is that developers must give inline views a width and a height via the style prop. Here's a screenshot illustrating this feature:

![image](https://cloud.githubusercontent.com/assets/199935/16635219/3fd4e5b8-4386-11e6-88b0-0b771ebd3a38.png)

Note that the changes to NativeViewHierarchyOptimizer were broken up across a few commits to make the diffs easier to understand.

Previously, inline Images were supported via FrescoBasedReactTextInlineImageSpan. To get support for nesting views within Text, we create one special kind of span per inline view. This span is called TextInlineViewPlaceholderSpan. It is the same size as the inline view. Its job is just to occupy space -- it doesn't render any visual. After the text is rendered, we query the Android Layout object associated with the TextView to find out where it has positioned each TextInlineViewPlaceholderSpan. We then position the views to be at those locations.

One tricky aspect of the implementation is that the Text component needs to be able to render native children (the inline views) but the Android TextView cannot have children. This is solved by having the native parent of the ReactTextView also host the inline views. Implementation-wise, this was accomplished by extending the NativeViewHierarchyOptimizer to handle this case. The optimizer now handles these cases:
- Node is not in the native tree. An ancestor must host its children.
- Node is in the native tree and it can host its own children.
- (new) Node is in the native tree but it cannot host its own children. An ancestor must host both this node and its children.

**Limitation: Clipping**

If Text's height/width is small such that an inline view doesn't completely fit, the inline view may still be fully visible due to hoisting (the inline view isn't actually parented to the Text which has the limited size. It is parented to an ancestor which may have a different clipping rectangle.). Prior to this change, layout-only views had a similar limitation.

**Test plan (required)**

This change was tested with UIExplorer and a small test app and it's being used in my team's app.

**Outstanding Issue: Using Text and Image within an inline view**

Currently, you cannot use Text or Image components within an inline view. This is because the React context flag, `isInAParentText`, indicates that you're still inside of a Text component and these components use their virtual node representation rather than their normal representation.

The Text component sets `isInAParentText` to `true`. No components ever set `isInAParentText` to `false`.

I need some help with fixing this. One option I was considering was setting `isInAParentText` to `false` for all other native components. This could be done by editing the native component creation primitives such as `requireNativeComponent` and `createReactNativeComponentClass` to provide the appropriate context. However, when I tried this, it didn’t work and it wasn’t clear to me whether or not context works on native components. Thoughts?

Adam Comella
Microsoft Corp.

facebook-github-bot-8 commented:
By analyzing the blame information on this pull request, we identified @dmmiller and @korDen to be potential reviewers.

dmmiller commented:
This is a lot, I'll look tomorrow (Friday)

dmmiller commented:
I'm looking through this.  It might actually be better to have this as separate PRs.  When we import them, all the commits get squashed into one request.  Would probably be easier to get through.

dmmiller commented on ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextShadowNode.java:
```diff
@@ -348,6 +355,15 @@ public ReactTextShadowNode(boolean isVirtual) {
 
+  // Returns a line height which takes into account the requested line height
+  // and the height of the inline images.
+  public float getAffectiveLineHeight() {
```
should be effectiveLineHeight

rigdern replied:
Fixed

dmmiller commented:
The first commit seems good (modulo the one thing inline)

dmmiller commented:
The second commit is solid.  Nice refactoring.

dmmiller commented:
Ditto for third.

rigdern commented:
@dmmiller I haven't seen any updates in a few days. Are you still reviewing this or are you waiting for me to take some action?

dmmiller commented:
@rigdern Sorry, was on holiday last week.  I thought I noted that here, but I didn't.  I have looked at the first three commits.  I'd actually prefer them to go in separately because we actually squash any PR into one commit so we would end up with a giant one.  They seem independent and good cleanup and I can quickly stamp them since I've already looked at them. Then you can rebase the fourth and just that one can be looked at in more detail.  Again sorry for the delay, I thought I had mentioned I was on holiday but forgot.

rigdern commented:
@dmmiller Okay, here's what I did. I addressed your feedback, rebased all of my commits onto the latest master, and tested the changes. Then I split this into 2 other PRs:
  - #8907: Line height fix for inline images
  - #8908: Refactoring of HierarchyOptimizer

This PR still includes the other 3 commits (because it builds on them) but it has been rebased onto the latest master. You can review the last commit (Enable views to be nested within \<Text\>). After the other PRs are accepted, I will rebase this on master and remove the other 3 commits from this PR.

In addition to reviewing the last commit, a couple of other things still need to happen:
  1. Merge PR https://github.com/facebook/css-layout/pull/202 into `css-layout` and then import it into `react-native`.
  2. Solve the outstanding issue for using Text and Image within an inline view. See my initial PR description for details.

dmmiller commented:
Accepted those two PRs

I don't know about css, @emilsjolander is the right person for that, and I assume you are working with him on the css PR.

emilsjolander commented:
Yup, looking at merging css-layout end of this or early next week.

rigdern commented:
@dmmiller Thanks for the quick turnaround on merging those 2 PRs. I rebased and this PR now only has 1 commit.

dmmiller commented:
@astreet Do you mind looking at some of the NativeHierarchyManager and UIImplementation pieces?

astreet commented:
I'm concerned about the complexity this adds on Android, especially with adding the notion of a layout parent in addition to a normal parent in the shadow tree. I think we might end up seeing some odd behavior with this reparenting as well (e.g. in iteration of focus and accessibility regions, touch handling, etc).

Can you give me an example of what you're trying to build? Is it possibly something that could be done by hosting inline drawables (instead of just fresco images?)

If we want to go with something like the current approach, I'd probably be happier if we explicitly created a new ViewGroup parent to hold the Text and the inline View in the case where we have an inline View. I think this would let us keep from having to do reparenting (I know we do reparenting now in the optimizer, but this reparenting is very simple and only gets rid of non-drawing FrameLayouts). We'd probably want to drive this logic from JS (e.g. in the render() in Text.js, we might have to check if any of the children are not strings/Text/RawTexts). Note that hosting in a special parent wouldn't automatically get rid of the iteration of focus/accessibility/touch issues, but it'd allow us to fix and contain them in a more isolated setting. Thoughts?

rigdern commented:
> I think we might end up seeing some odd behavior with this reparenting as well (e.g. in iteration of focus and accessibility regions, touch handling, etc).

Can you elaborate on the odd behaviors you're afraid we might see?

> Can you give me an example of what you're trying to build? Is it possibly something that could be done by hosting inline drawables (instead of just fresco images?)

We're trying to render [animated sprites](https://jsfiddle.net/simurai/CGmCe/) nested inside of `Text` components.

I suspect that this could be implemented as a Drawable but I haven't tried. If RN Android supported custom inline Drawables, would this require the app developer to write Android-specific Java code? One nice characteristic of inline views is that it enables developers to express their inline visuals using cross-platform JavaScript.

[RN for iOS supports inline views](https://github.com/facebook/react-native/pull/7304) so it'd be nice to have feature parity in the Android version.

> We'd probably want to drive this logic from JS (e.g. in the render() in Text.js, we might have to check if any of the children are not strings/Text/RawTexts).

I considered something like this in my implementation, however I wasn't sure how to determine through JavaScript whether or not the Text contained any non-strings/Text/RawTexts. The problem I faced is that `this.props.children` contains a mixture of native primitives (e.g. `View`, `Text`, `ScrollView`) and components (subclasses of `React.Component`). When you encounter a `React.Component`, you want to know what native primitives it'll render down into. I don't know of any easy way of getting at this information in JavaScript.

> If we want to go with something like the current approach, I'd probably be happier if we explicitly created a new ViewGroup parent to hold the Text and the inline View in the case where we have an inline View.

If you agree that we should have support for inline views on RN Android, I can look into this alternative solution.

astreet commented:
>Can you elaborate on the odd behaviors you're afraid we might see?

In the case of adding arbitrary view groups, it's possible that event capture/bubbling might not work correctly: e.g. on the native side, both capture and bubbling would not go through the text 'parent' (since it's no longer the parent). This could be an issue depending on what pointer-events you have set. Enumerating tappable views for accessibility and focus would also be in an incorrect order as it would traverse the tappable views in the text, skipping over the view that is no longer rendered within the text, separate from traversing the views that are children of the container that view is now being owned by on the native side.

>I suspect that this could be implemented as a Drawable but I haven't tried. If RN Android supported custom inline Drawables, would this require the app developer to write Android-specific Java code? One nice characteristic of inline views is that it enables developers to express their inline visuals using cross-platform JavaScript.

Yeah, that's what I figured: I would be interested to hear if there are actually non-Drawable usecases for this (if anyone is following along and has an example of something they want to implement, please chime in).

I researched some yesterday about how I would implement this natively (e.g. if I wanted to host an arbitrary view inline within text) and it really doesn't seem very easy even in native (I'm thinking I would probably create a custom ViewGroup that can also render arbitrary text spans interleaved with non-Text children, and even that would need custom logic for touch handling, accessibility, etc). On top of that, I couldn't find anyone asking on SO how to host things inline in text outside of rendering drawables, so I don't think it's a very common use case. If all you need is drawable support, I would just add that.

As far as 'custom inline drawables', I think we'd still just expose image drawables, but maybe we can add support for animated images/animated inline images?

>I considered something like this in my implementation, however I wasn't sure how to determine through JavaScript whether or not the Text contained any non-strings/Text/RawTexts. The problem I faced is that this.props.children contains a mixture of native primitives (e.g. View, Text, ScrollView) and components (subclasses of React.Component). When you encounter a React.Component, you want to know what native primitives it'll render down into. I don't know of any easy way of getting at this information in JavaScript.

Hmm, this is a tough one. I wonder if @sahrens might have a good idea? I'll link this to him

rigdern commented:
I know you stated otherwise, but it seems to me that representing the `Text` component as a `ViewGroup` which contains the `TextView` and the inline views would mitigate the eventing problem. We'd just have to ensure that the `Text` component's event handlers are placed on the `ViewGroup` rather than on the `TextView`. It seems like this design would be similar to the case where you create an [`Image` component which has children](https://github.com/facebook/react-native/blob/1ab4b2a7924767364d54bc96ed99c93a3916cf88/Libraries/Image/Image.android.js#L269-L272).

If I understand your accessibility concern correctly, you are concerned that you will not be able to enumerate tappables in the proper order which the user has placed inside of the `Text` component. The problem being that some tappables will be inline views which get placed as siblings of the `TextView` in the native view hierarchy whereas other tappables will remain as children of the `TextView`. What views fall into the latter case? I can only think of inline images but I don't think inline images can be tappable.

For the option of supporting 'custom inline drawables' in RN Android, do you imagine developers would write some Java code to implement their custom Drawable and then get that hooked into RN similar to how custom modules and view managers are hooked into RN today?

astreet commented:
>I know you stated otherwise, but it seems to me that representing the Text component as a ViewGroup which contains the TextView and the inline views would mitigate the eventing problem. We'd just have to ensure that the Text component's event handlers are placed on the ViewGroup rather than on the TextView. It seems like this design would be similar to the case where you create an Image component which has children.

Yeah, I agree creating a separate explicit ViewGroup to mount the child Views in would be preferable and would mitigate most of my concerns around event handling, etc. Even then, though, I still probably fall in the camp of making a simple change to support what we know we need than a more invasive change that we need to maintain going forward in order to support a general case whose value isn't clear.

>What views fall into the latter case?

I was thinking about tappable text Spans that (e.g. you can have child `<Text>` components that have click handlers, e.g. to support inline hyperlinks)

>For the option of supporting 'custom inline drawables' in RN Android, do you imagine developers would write some Java code to implement their custom Drawable and then get that hooked into RN similar to how custom modules and view managers are hooked into RN today?

I'm actually proposing not supporting arbitrary inline drawables but instead just explicitly adding support for inline animated images in a similar way to how we do images now. I'm having trouble thinking of what other kinds of drawables people would be interested in embedding in text that we couldn't just add simple explicit support for.

rigdern commented:
That's a good point about hyperlinks being clickable inside of `<Text>` components. Did you have any solutions in mind for providing a proper enumeration order for inline views and inline hyperlinks for accessibility/focus purposes?

Just building support for animated sprites into React Native is an interesting proposal. At the moment, I can't think of any other inline view scenarios. For our app, we built an animated sprite component which we nest inside of `<Text>`. Here are some of the features we use and would need support for in whatever solution we end up with here:
  - Event to find out when the animation has been loaded and is ready to be played.
  - Ability to render a static image while the animation is loading.
  - Ability to start the animation on any frame.
  - Ability to control when the animation starts so it can be coordinated with other things.
  - Ability to pause the animation and resume it later.
  - Ability to provide an optimized sprite sheet which contains no duplicated frames. To use this feature, you also need to provide an array of indices into the sprite sheet representing the order the frames should be shown in.

I had an e-mail thread with @vjeux a while ago before submitting this PR about building support into React Native for animated sprites vs inline views. Christopher, do you have any thoughts on this topic? Did you have any inline view scenarios in mind besides animated images?

aaronschubert0 commented:
@rigdern @astreet I've been following this thread, thanks so much for adding support for this. The scenario that we would be using inline views is to have decorated hyperlinks (as per the attached image) The user would be able to tap on these (TouchableOpacity) and then we can animate this view.

So I think a generalised solution as is the case on iOS would be most beneficial for us in the short term.

![screen shot 2016-08-04 at 09 03 35](https://cloud.githubusercontent.com/assets/9281756/17394689/6a2125b8-5a22-11e6-966b-e626186fbdb3.png)

rigdern commented:
@astreet 

> My main point is I think isolation will make it easier and less bug-prone.

I'm fine with changing the inline view implementation to wrap each `Text` component which contains inline views with another `View`.

I thought more about this problem and I'm glad that you agree with my requirements for an animated sprite component but I'm concerned this solution might be too limiting for us. Later, we may have new requirements for the animated sprite component and you may not agree with them. At the very least, I think we should have an escape hatch to be able to implement our own custom Drawables for use with React Native's Text component. From an app developer's perspective, support for inline views would be even better because it offers more flexibility and the flexibility can be leveraged from JavaScript rather than native code.

Let's summarize the options for supporting inline visuals:

1. **Support inline animated sprites.** The main downside is this solution offers no flexibility to the app developer if they have different requirements for the sprite component or they want a different kind of inline visual. The benefits include:
  - Simple implementation that covers the known use case.
2. **Support custom Drawables.** The main downside is you need to write platform-specific code to leverage this solution. The benefits include:
  - Flexibility to implement custom inline visuals.
3. **Support inline views on Android.** The focus and accessibility issues you listed with this approach seem solvable. The main downside is the extra complexity it adds to the code base. The benefits include:
  - Flexibility to implement inline visuals using cross platform JavaScript code.
  - Parity with the iOS implementation of React Native.

My preference is for (3). It's nice to have flexibility in a generic UI framework like React Native. It'll enable our app to tackle future problems such as changes in the requirements for animated sprites and any need for a new inline visual.

astreet commented:
Yeah, no argument from me about inline views providing more potential value :) It's all about what we have to pay for them. Given we don't have specific use cases and that this isn't asked for outside of images, I think my litmus test would be whether we can add support for them without adding new code paths we have to maintain within the core layouting and UI code.

Re: new requirements for sprites we might not agree with: this is getting theoretical, but if there are things you want to do that we don't want to support, there's always the possibility of writing your own view manager for your own inline image/drawable subclass.

rigdern commented:
> there's always the possibility of writing your own view manager for your own inline image/drawable subclass.

In order for this to work, we need to implement either option (2) custom Drawables or (3) inline views.

> I think my litmus test would be whether we can add support for them without adding new code paths we have to maintain within the core layouting and UI code.

I can look into doing an implementation of inline views in which Text components which have inline views get wrapped by a View. That way, we won't have to make changes to the NativeViewHierarchyOptimizer and might be able to keep the changes contained to the Text component.

@astreet, can you find somebody who can help give advice on the following problems (they might look familiar because they were mentioned earlier in this issue)?:
  1. **Using Text and Image within an inline view.** Currently, you cannot use Text or Image components within an inline view. This is because the React context flag, `isInAParentText`, indicates that you're still inside of a Text component and these components use their virtual node representation rather than their normal representation. The Text component sets `isInAParentText` to `true`. No components ever set `isInAParentText` to `false`. I need some help with fixing this. One option I was considering was setting `isInAParentText` to `false` for all other native components. This could be done by editing the native component creation primitives such as `requireNativeComponent` and `createReactNativeComponentClass` to provide the appropriate context. However, when I tried this, it didn’t work and it wasn’t clear to me whether or not context works on native components.
  2. **Inspecting native primitives in JavaScript.** Can we have the JavaScript logic for the `render` function of Text components output different things depending on whether or not all of its children are strings/Texts/RawTexts? The problem I faced is that `this.props.children` contains a mixture of native primitives (e.g. `View`, `Text`, `ScrollView`) and components (subclasses of `React.Component`). When you encounter a `React.Component`, you want to know what native primitives it'll render down into. I don't know of any easy way of getting at this information in JavaScript.

facebook-github-bot-2 commented:
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @emilsjolander as a potential reviewer. Could you take a look please or cc someone with more context?

astreet commented:
Sorry about the delay on this, I was out:

>In order for this to work, we need to implement either option (2) custom Drawables or (3) inline views.

When we're talking about adding specifically image-rendering functionality that we don't think is necessarily useful or worth 'paying for' to the core library, custom Drawables doesn't seem bad to me. Currently, you would be able to create a custom ViewManager (a la https://github.com/facebook/react-native/blob/1926fdf156e077ad8b0c73505c6447d1a16687f8/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageViewManager.java) and extend ReactTextInlineImageShadowNode (see https://github.com/facebook/react-native/search?utf8=%E2%9C%93&q=ReactTextInlineImageShadowNode&type=Code)

Given the difficulties you're listing (which I don't think are easily solvable), at this point I feel pretty confident that adding animated inline image support specifically is the right call here.

facebook-github-bot-8 commented:
@rigdern updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8619/files/249ef5061001fb364bda49c733d1300b3c590ad1..c74001ddf0332a3593c1048064d5617efea725dc)

facebook-github-bot-4 commented:
@rigdern updated the pull request - [view changes](https://github.com/facebook/react-native/pull/8619/files/c74001ddf0332a3593c1048064d5617efea725dc..249ef5061001fb364bda49c733d1300b3c590ad1)

rigdern commented:
I accidentally pushed to this branch. There's nothing new to review yet.

huangmr commented:
any update on this PR please?

astreet commented:
@huangmr: What are you trying to achieve? Would be good to know what other people are trying to do with this patch.

mkonicek commented:
@rigdern I'm curious - did the iOS implementation end up getting landed again after being reverted? https://github.com/facebook/react-native/pull/7304

rigdern commented:
@mkonicek Yes, 486dbe4e8f4e3710c258e22af00b2aa7155ef2f3 has been in RN since 0.28.0.

hramos commented:
@rigdern do you still plan to add more commits to this PR or is it ready for review?

rigdern commented:
@hramos, last time @astreet reviewed this, he expressed concerns with the current approach and suggested a couple of other approaches. Unless opinions have changed on my original approach, I have to try out one of the other approaches and update the PR before it can be reviewed. Because investigating the other approaches will be time consuming and the current implementation is working well for my team's app, I haven't had the chance to revisit this PR yet.

lacker commented:
It sounds like you might just never revisit this PR. Should we close it?

rigdern commented:
@lacker I intend on coming back to this at some point. My team needs this feature and is currently using this PR in our app. We want this feature to be part of the official RN repo. Revisiting this PR will be time consuming so I can't justify spending time on it right now given the priorities of my other work.

If you'd like, I can close the PR and repoen it or open a new one when I revisit it. What do you think?

lacker commented:
That sounds good. Let's open a new one when you revisit - I think that will just help us organizationally get a handle on our pull request backlog. Thanks for understanding!
